test/torchaudio_unittest/backend/sox_io/info_test.py

```diff
@@ -221,6 +221,23 @@ def test_alaw(self):
         assert info.bits_per_sample == 8
         assert info.encoding == "ALAW"
 
+    def test_gsm(self):
+        """`sox_io_backend.info` can check gsm file correctly"""
+        duration = 1
+        num_channels = 1
+        sample_rate = 8000
+        path = self.get_temp_path('data.wav')
```

Comment:
Can you use 'gsm' as extension? The encoding is provided separately and correctly, but at a glance, it looks like the test is wrong.

Reply:
Hello @mthrok, if I do that, then the `info.num_frames` assert fails because its value is 0, instead of 8000.
Also, I noticed that there's a small bug in test/torchaudio_unittest/common_utils/sox_utils.py, which is producing a warning, as in the example below. I added a PR for it (#1287), but it doesn't fix the issue of `info.num_frames` being 0.
Please confirm if that field in the header is supposed to be 0, for some reason. 

```
>       assert info.num_frames == sample_rate * duration
E       AssertionError: assert 0 == 8000
E         -0
E         +8000

info_test.py:251: AssertionError

----------------------------------------------------------------------------- Captured stderr call ------------------------------------------------------------------------------
sox -V3 --no-dither -R --bits 16 --rate 8000 --null --channels 1 --bits 16 --encoding gsm /tmp/tmp08b_xoao/torchaudio_unittest.backend.sox_io.info_test.TestInfo.test_gsm/data.gsm synth 1 sawtooth 1
sox:      SoX v14.4.2

Input File     : '' (null)
Channels       : 1
Sample Rate    : 8000
Precision      : 16-bit
Endian Type    : little
Reverse Nibbles: no
Reverse Bits   : no

sox WARN formats: gsm can't encode GSM to 16-bit

Output File    : '/tmp/tmp08b_xoao/torchaudio_unittest.backend.sox_io.info_test.TestInfo.test_gsm/data.gsm'
Channels       : 1
Sample Rate    : 8000
Precision      : 16-bit
Sample Encoding: GSM
Comment        : 'Processed by SoX'

sox INFO sox: effects chain: input         8000Hz  1 channels
sox INFO sox: effects chain: synth         8000Hz  1 channels
sox INFO sox: effects chain: output        8000Hz  1 channels
=================================================================== 1 failed, 215 deselected in 0.84 seconds ====================================================================
```

Reply:
libsox is writing the `length` field in `sox_signalinfo_t` as 0 for `gsm`, which is why `get_info_file` provides 0 value for `AudioMetaData::num_frames`.

Reply:
In short, you can remove the check for `num_frames`. `soxi` does not report the number of samples either.

<details><summary>example</summary>

```
% sox -V3 --no-dither -R --rate 8000 --null --channels 1 foo.gsm synth 1 sawtooth 1
sox:      SoX v

Input File     : '' (null)
Channels       : 1
Sample Rate    : 8000
Precision      : 32-bit


Output File    : 'foo.gsm'
Channels       : 1
Sample Rate    : 8000
Precision      : 16-bit
Sample Encoding: GSM
Comment        : 'Processed by SoX'

sox INFO sox: effects chain: input         8000Hz  1 channels
sox INFO sox: effects chain: synth         8000Hz  1 channels
sox INFO sox: effects chain: output        8000Hz  1 channels

% soxi foo.gsm

Input File     : 'foo.gsm'
Channels       : 1
Sample Rate    : 8000
Precision      : 16-bit
Sample Encoding: GSM
```
</details>

I am not an expert in audio formats and/or libsox but here is my educated guess. When you use `wav` as a file extension, the generated file is WAV format with GSM encoding. So probably the header retains the information about the number of samples. When you use `gsm` extension, the format becomes GSM. I still cannot find a document to validate this hypothesis, so not entirely sure, but with this format, number of samples are not found in header. This is also the case for other streaming formats. MP3 and flac when streamed reports 0 frames too.

Reply:
Thank you! I think your hypothesis is correct! I had changed the file extension to `gsm` in the morning.
There's another issue, BTW. GSM supports up to 16 channels, but if the num_channels parameter is more than 1,
even then soxi reports 1 channel.
I also tested it manually.

Reply:
> GSM supports up to 16 channels, but if the num_channels parameter is more than 1, even then soxi reports 1 channel.

I think the issue lies in the generation process. `ffmpeg` also reports one channel (with the duration of num_channels second). So It is not correct to say that torchaudio's `info` function always reports `num_channels==1`. Since we do not have a way to test, we can just test the case where `num_channels=1`.

```
% sox -V3 --no-dither -R --rate 8000 --null --channels 13 foo.gsm synth 1 sawtooth 1
sox:      SoX v

Input File     : '' (null)
Channels       : 1
Sample Rate    : 8000
Precision      : 32-bit


Output File    : 'foo.gsm'
Channels       : 13
Sample Rate    : 8000
Precision      : 16-bit
Sample Encoding: GSM
Comment        : 'Processed by SoX'

sox INFO sox: effects chain: input         8000Hz  1 channels
sox INFO sox: effects chain: synth         8000Hz  1 channels
sox INFO sox: effects chain: channels      8000Hz 13 channels
sox INFO sox: effects chain: output        8000Hz 13 channels
% ffprobe foo.gsm
ffprobe version 4.3.1 Copyright (c) 2007-2020 the FFmpeg developers
  built with Apple clang version 12.0.0 (clang-1200.0.32.27)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/4.3.1_4 --enable-shared --enable-pthreads --enable-version3 --enable-avresample --cc=clang --host-cflags= --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libbluray --enable-libdav1d --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librtmp --enable-libspeex --enable-libsoxr --enable-videotoolbox --disable-libjack --disable-indev=jack
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100
[gsm @ 0x7ffa35013c00] Estimating duration from bitrate, this may be inaccurate
Input #0, gsm, from 'foo.gsm':
  Duration: 00:00:13.00, start: 0.000000, bitrate: 13 kb/s
    Stream #0:0: Audio: gsm, 8000 Hz, mono, s16, 13 kb/s
```

Reply:
Thanks a lot for your help! 
I'll change the test to test for just one channel.